// 3d-tiles-renderer/build/index.plugins.js
import { EnforceNonZeroErrorPlugin as s, ImplicitTilingPlugin as i } from "./index.core-plugins.js";
import { C as n, G as o, Q as g, z as u } from "./QuantizedMeshLoaderBase-018lzCXI.js";
import { m as t, B as T, C as P, k as d, D as m, o as C, e as M, a as p, g as O, f as S, h, G, l as y, I as L, L as R, O as v, Q as x, n as F, R as I, S as W, j as c, q as B, T as E, d as f, c as Z, b, U as z, u as A, W as D, s as Q, t as U, i as X, r as N, X as Y, p as j } from "./WMSCapabilitiesLoader-Ca3IzEbJ.js";
export {
  t as BaseRegion,
  T as BatchedTilesPlugin,
  n as CesiumIonAuth,
  P as CesiumIonAuthPlugin,
  d as CesiumIonOverlay,
  m as DebugTilesPlugin,
  C as DeepZoomImagePlugin,
  s as EnforceNonZeroErrorPlugin,
  M as GLTFCesiumRTCExtension,
  p as GLTFExtensionsPlugin,
  O as GLTFMeshFeaturesExtension,
  S as GLTFStructuralMetadataExtension,
  h as GeoJSONOverlay,
  o as GoogleCloudAuth,
  G as GoogleCloudAuthPlugin,
  y as GoogleMapsOverlay,
  L as ImageOverlayPlugin,
  i as ImplicitTilingPlugin,
  R as LoadRegionPlugin,
  v as OBBRegion,
  g as QuantizedMeshLoaderBase,
  x as QuantizedMeshPlugin,
  F as RayRegion,
  I as ReorientationPlugin,
  W as SphereRegion,
  c as TMSTilesOverlay,
  B as TMSTilesPlugin,
  E as TileCompressionPlugin,
  f as TileFlatteningPlugin,
  Z as TilesFadePlugin,
  b as UnloadTilesPlugin,
  z as UpdateOnChangePlugin,
  A as WMSCapabilitiesLoader,
  D as WMSTilesOverlay,
  Q as WMSTilesPlugin,
  U as WMTSCapabilitiesLoader,
  X as WMTSTilesOverlay,
  N as WMTSTilesPlugin,
  Y as XYZTilesOverlay,
  j as XYZTilesPlugin,
  u as zigZagDecode
};
